lsp-mode.el

```diff
@@ -7871,19 +7868,13 @@ CALLBACK is the status callback passed by Flycheck."
 
 (defun lsp--flycheck-report ()
   "This callback is invoked when new diagnostics are received
-from the language server. Invoke flycheck-buffer to update the
-display of errors if flycheck-mode is on and we are live
-reporting or we are in save-mode and the buffer is not modified."
-  (cond
-   ;; do nothing
-   ((and (not lsp-flycheck-live-reporting)
-         (not (and (memq 'save flycheck-check-syntax-automatically)
-                   (not (buffer-modified-p))))))
-   ;; visible and not inhibit hooks - refresh right away
-   ((and (not lsp-inhibit-lsp-hooks) (lsp--buffer-visible?))
-    (flycheck-buffer))
-   ;; not visible or inhibit hooks - schedule refresh
-   (t (add-hook 'lsp-on-idle-hook #'lsp--flycheck-buffer nil t))))
+from the language server."
+  (when (not lsp--diagnostics-modified?)
+    (setq lsp--diagnostics-modified? t)
+    (when (memq 'idle-change flycheck-check-syntax-automatically)
+      ;; make sure diagnostics are published even if the diagnostics
+      ;; have been received after idle-change has been triggered
+      (add-hook 'lsp-on-idle-hook #'lsp--flycheck-buffer nil t))))
```

Comment:
there is one more thing that we should do here, consider the following: User edits -> flycheck idle passes and it does not report the errors -> server is slow and takes 5 seconds to calculate and send diagnostics -> lsp-mode receives the messages and it will show them on in lsp-on-idle -> but the user does not perform any action, waiting for the errors to show up.

If the user does not do anything, the error won't pop up - I think you should call `(lsp--idle-reschedule (current-buffer))` here.